core/src/main/java/org/apache/calcite/sql/SqlUpdate.java

```diff
@@ -160,12 +166,104 @@ public void setSourceSelect(SqlSelect sourceSelect) {
         writer.startList(SqlWriter.FrameTypeEnum.SELECT, "UPDATE", "");
     final int opLeft = getOperator().getLeftPrec();
     final int opRight = getOperator().getRightPrec();
+
+    List<SqlNode> sources = new ArrayList<>();
+
+    if (sourceSelect != null && sourceSelect.from != null) {
+      SqlJoin join = getJoinFromSourceSelect();
+      if (join != null) {
+        setTargetAndSources(join, sources);
+      }
+    }
+
     targetTable.unparse(writer, opLeft, opRight);
+
     SqlIdentifier alias = this.alias;
     if (alias != null) {
       writer.keyword("AS");
       alias.unparse(writer, opLeft, opRight);
     }
+
+    unparseSet(writer, opLeft, opRight);
+    if (!sources.isEmpty()) {
+      unparseUpdateSources(sources, writer, opLeft, opRight);
+    }
+    unparseUpdateCondition(writer, opLeft, opRight);
+
+    writer.endList(frame);
+  }
+
+  private SqlIdentifier getFromAlias() {
+    if (sourceSelect != null && sourceSelect.from != null) {
+      if (sourceSelect.from instanceof SqlBasicCall && sourceSelect.from.getKind() == SqlKind.AS) {
+        return (SqlIdentifier) ((SqlBasicCall) sourceSelect.from).operands[1];
+      }
+    }
+    return null;
+  }
+
+  private SqlJoin getJoinFromSourceSelect() {
+    SqlJoin join = null;
+    if (sourceSelect.from instanceof SqlBasicCall && sourceSelect.from.getKind() == SqlKind.AS
+        && ((SqlBasicCall) sourceSelect.from).operands[0] instanceof SqlJoin) {
+      join = (SqlJoin) ((SqlBasicCall) sourceSelect.from).operands[0];
+    } else if (sourceSelect.from instanceof SqlJoin) {
+      join = (SqlJoin) sourceSelect.from;
+    }
+    return join;
+  }
+
+  private void setTargetAndSources(SqlNode node, List<SqlNode> sources) {
+    switch (node.getKind()) {
+    case JOIN:
+      setTargetAndSources((SqlJoin) node, sources);
+      break;
+    case IDENTIFIER:
+      setTargetAndSources((SqlIdentifier) node, sources);
+      break;
+    case AS:
+      setTargetAndSources((SqlBasicCall) node, sources);
+      break;
+    case WITH:
+      setTargetAndSources((SqlWith) node, sources);
+      break;
+    }
+  }
+
+  private void setTargetAndSources(SqlJoin node, List<SqlNode> sources) {
+    setTargetAndSources(node.right, sources);
+    setTargetAndSources(node.left, sources);
+  }
+
+  private void setTargetAndSources(SqlIdentifier node, List<SqlNode> sources) {
+    if (!isTargetTable(node)) {
+      sources.add(node);
+    }
+  }
+
+  private void setTargetAndSources(SqlBasicCall node, List<SqlNode> sources) {
+    if (isTargetTable(node)) {
+      setTargetTable(node);
+    } else {
+      sources.add(node);
+    }
+  }
+
+  private boolean isTargetTable(SqlNode node) {
```

Comment:
Can you document about the 3 cases handled as part of this method.

Reply:
Done